### ISSUE TITLE
Stack overflow: infinite loop when clearing inventory

### ISSUE THREAD

**Issue by mattysweeps** [SpongePowered/SpongeForge#1262]
When my plugin tried to clear an inventory (I believe a chest), this happened:
```java
java.lang.StackOverflowError
	at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81) ~[DefaultInventoryFabric.class:?]
	at net.minecraft.tileentity.TileEntityChest.clear(TileEntityChest.java:632) ~[TileEntityChest.class:?]
	at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81) ~[DefaultInventoryFabric.class:?]
	at net.minecraft.tileentity.TileEntityChest.clear(TileEntityChest.java:632) ~[TileEntityChest.class:?]
	at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81) ~[DefaultInventoryFabric.class:?]
	at net.minecraft.tileentity.TileEntityChest.clear(TileEntityChest.java:632) ~[TileEntityChest.class:?]
```
full stack trace here: https://gist.github.com/m0pt0pmatt/87a7895c297aede1ea5b54fae4822738

It's hard to know for sure, but digging through my code I think this is the code which started the infinite loop:
```java
        String worldName = getOrThrow(survivalGame.getConfig().getWorldName(), CommandKeys.WORLD_NAME);
        World world = getOrThrow(Sponge.getServer().getWorld(worldName), CommandKeys.WORLD_NAME);
        Integer chestMidpoint = getOrThrow(survivalGame.getConfig().getChestMidpoint(), CommandKeys.CHEST_MIDPOINT);
        Integer chestRange = getOrThrow(survivalGame.getConfig().getChestRange(), CommandKeys.CHEST_RANGE);

        world.getTileEntities().forEach(tileEntity -> {

            if (tileEntity instanceof Chest) {
                Chest chest = (Chest) tileEntity;
                chest.getInventory().clear();

                if (!survivalGame.getConfig().getItemConfig().getItems().isEmpty()) {

                    double itemCount = (chestMidpoint + ((RANDOM.nextDouble() * chestRange) * (RANDOM.nextDouble() > 0.5 ? 1 : -1)));
                    for (int i = 0; i < itemCount; i++) {
                        ItemStackSnapshot stackSnapshot = survivalGame.getConfig().getItemConfig().getItems().get(RANDOM.nextInt(survivalGame.getConfig().getItemConfig().getItems().size()));
                        chest.getInventory().offer(stackSnapshot.createStack());
                    }
                }
            }
        });
```

I'm using the bleeding branch of SpongeForge, and it's up to date.

I'm going to debug my server for more information.

**Comment by mattysweeps** [comment]
Yup, that's the call which starts the infinite loop. Confirmed it in the debugger.

Looks like the culprit is the mixin TraitInventoryAdapter.

After taking a thread dump:
```java
"Server thread@4767" prio=5 tid=0x14 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	 blocks Netty Server IO #2@10710
	 blocks Netty Server IO #1@10016
	  at net.minecraft.tileentity.TileEntityChest.clear(TraitInventoryAdapter.java:132)
	  at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81)
	  at net.minecraft.tileentity.TileEntityChest.clear(TraitInventoryAdapter.java:132)
	  at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81)
	  at net.minecraft.tileentity.TileEntityChest.clear(TraitInventoryAdapter.java:132)
	  at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81)
	  at net.minecraft.tileentity.TileEntityChest.clear(TraitInventoryAdapter.java:132)
	  at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81)
	  at net.minecraft.tileentity.TileEntityChest.clear(TraitInventoryAdapter.java:132)
	  at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81)
	  at net.minecraft.tileentity.TileEntityChest.clear(TraitInventoryAdapter.java:132)
	  at org.spongepowered.common.item.inventory.lens.impl.fabric.DefaultInventoryFabric.clear(DefaultInventoryFabric.java:81)
	  at net.minecraft.tileentity.TileEntityChest.clear(TraitInventoryAdapter.java:132)
	  at io.github.m0pt0pmatt.survivalgames.task.FillChestsTask.lambda$execute$0(FillChestsTask.java:57)
	  at io.github.m0pt0pmatt.survivalgames.task.FillChestsTask$$Lambda$557.50707023.accept(Unknown Source:-1)
	  at java.util.ArrayList.forEach(ArrayList.java:1249)
	  at io.github.m0pt0pmatt.survivalgames.task.FillChestsTask.execute(FillChestsTask.java:53)
	  at io.github.m0pt0pmatt.survivalgames.game.SurvivalGameStateManager.executeTasks(SurvivalGameStateManager.java:198)
	  at io.github.m0pt0pmatt.survivalgames.game.SurvivalGameStateManager.start(SurvivalGameStateManager.java:116)
	  at io.github.m0pt0pmatt.survivalgames.command.executor.state.StartSurvivalGameCommand$$Lambda$373.49469441.accept(Unknown Source:-1)
	  at io.github.m0pt0pmatt.survivalgames.command.executor.state.AbstractStateCommand.execute(AbstractStateCommand.java:80)
	  at io.github.m0pt0pmatt.survivalgames.command.executor.state.StartSurvivalGameCommand.execute(StartSurvivalGameCommand.java:34)
	  at org.spongepowered.api.command.args.ChildCommandElementExecutor.execute(ChildCommandElementExecutor.java:179)
	  at org.spongepowered.api.command.spec.CommandSpec.process(CommandSpec.java:351)
	  at org.spongepowered.api.command.dispatcher.SimpleDispatcher.process(SimpleDispatcher.java:333)
	  at org.spongepowered.common.command.SpongeCommandManager.process(SpongeCommandManager.java:295)
	  at net.minecraft.command.ServerCommandManager.executeCommand(MixinServerCommandManager.java:83)
	  at net.minecraft.network.NetHandlerPlayServer.handleSlashCommand(NetHandlerPlayServer.java:946)
	  at net.minecraft.network.NetHandlerPlayServer.processChatMessage(NetHandlerPlayServer.java:922)
	  at net.minecraft.network.play.client.CPacketChatMessage.processPacket(CPacketChatMessage.java:47)
	  at net.minecraft.network.play.client.CPacketChatMessage.processPacket(CPacketChatMessage.java:8)
	  at org.spongepowered.common.network.PacketUtil.lambda$onProcessPacket$0(PacketUtil.java:157)
	  at org.spongepowered.common.network.PacketUtil$$Lambda$442.1287563294.call(Unknown Source:-1)
	  at org.spongepowered.common.event.tracking.CauseTracker.switchToPhase(CauseTracker.java:156)
	  at org.spongepowered.common.network.PacketUtil.onProcessPacket(PacketUtil.java:156)
	  at net.minecraft.network.PacketThreadUtil$1.redirect$onProcessPacket$zji000(MixinPacketThreadUtil.java:39)
	  at net.minecraft.network.PacketThreadUtil$1.run(PacketThreadUtil.java:15)
	  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	  at net.minecraft.util.Util.runTask(Util.java:26)
	  at net.minecraft.server.MinecraftServer.updateTimeLightAndEntities(MinecraftServer.java:753)
	  - locked <0x2c7e> (a java.util.ArrayDeque)
	  at net.minecraft.server.dedicated.DedicatedServer.updateTimeLightAndEntities(DedicatedServer.java:402)
	  at net.minecraft.server.MinecraftServer.tick(MinecraftServer.java:698)
	  at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:547)
	  at java.lang.Thread.run(Thread.java:745)
```

It's strange, I didn't have this problem a few months ago, but both of the methods were implemented a while ago.

**Comment by mattysweeps** [comment]
Running the code on commit `e6ef93e52c39296c3ec8e5774cb1e483ec520bc9` everything works. The thread dump is:
```java
"Server thread@4557" prio=5 tid=0x14 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	 blocks Netty Server IO #3@9676
	  at net.minecraft.tileentity.TileEntityLockableLoot.fillWithLoot(TileEntityLockableLoot.java:53)
	  at net.minecraft.tileentity.TileEntityChest.clear(TileEntityChest.java:516)
	  at io.github.m0pt0pmatt.survivalgames.task.FillChestsTask.lambda$execute$0(FillChestsTask.java:57)
	  at io.github.m0pt0pmatt.survivalgames.task.FillChestsTask$$Lambda$566.411453813.accept(Unknown Source:-1)
	  at java.util.ArrayList.forEach(ArrayList.java:1249)
	  at io.github.m0pt0pmatt.survivalgames.task.FillChestsTask.execute(FillChestsTask.java:53)
	  at io.github.m0pt0pmatt.survivalgames.game.SurvivalGameStateManager.executeTasks(SurvivalGameStateManager.java:198)
	  at io.github.m0pt0pmatt.survivalgames.game.SurvivalGameStateManager.start(SurvivalGameStateManager.java:116)
	  at io.github.m0pt0pmatt.survivalgames.command.executor.state.StartSurvivalGameCommand$$Lambda$353.1180382385.accept(Unknown Source:-1)
	  at io.github.m0pt0pmatt.survivalgames.command.executor.state.AbstractStateCommand.execute(AbstractStateCommand.java:80)
	  at io.github.m0pt0pmatt.survivalgames.command.executor.state.StartSurvivalGameCommand.execute(StartSurvivalGameCommand.java:34)
	  at org.spongepowered.api.command.args.ChildCommandElementExecutor.execute(ChildCommandElementExecutor.java:179)
	  at org.spongepowered.api.command.spec.CommandSpec.process(CommandSpec.java:351)
	  at org.spongepowered.api.command.dispatcher.SimpleDispatcher.process(SimpleDispatcher.java:333)
	  at org.spongepowered.common.command.SpongeCommandManager.process(SpongeCommandManager.java:291)
	  at net.minecraft.command.ServerCommandManager.executeCommand(ServerCommandManager.java:82)
	  at net.minecraft.network.NetHandlerPlayServer.handleSlashCommand(NetHandlerPlayServer.java:955)
	  at net.minecraft.network.NetHandlerPlayServer.processChatMessage(NetHandlerPlayServer.java:931)
	  at net.minecraft.network.play.client.CPacketChatMessage.processPacket(CPacketChatMessage.java:47)
	  at net.minecraft.network.play.client.CPacketChatMessage.processPacket(CPacketChatMessage.java:8)
	  at org.spongepowered.common.network.PacketUtil.onProcessPacket(PacketUtil.java:122)
	  at net.minecraft.network.PacketThreadUtil$1.redirect$onProcessPacket$0(PacketThreadUtil.java:39)
	  at net.minecraft.network.PacketThreadUtil$1.run(PacketThreadUtil.java:15)
	  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	  at net.minecraft.util.Util.runTask(Util.java:25)
	  at net.minecraft.server.MinecraftServer.updateTimeLightAndEntities(MinecraftServer.java:742)
	  - locked <0x2b15> (a java.util.ArrayDeque)
	  at net.minecraft.server.dedicated.DedicatedServer.updateTimeLightAndEntities(DedicatedServer.java:408)
	  at net.minecraft.server.MinecraftServer.tick(MinecraftServer.java:687)
	  at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:536)
	  at java.lang.Thread.run(Thread.java:745)
```

so TraitInventoryAdapter wasn't being mixed into TileEntityChest?

**Comment by mattysweeps** [comment]
Just tried `67f5242db6f8cd9082efa66ff1d5330d39918c33` and it also didn't work. I'll keep looking

**Comment by Aaron1011** [comment]
This looks like an issue with the `@Intrinsic` for Inventory#clear not being skipped. I'll keep looking into it.

**Comment by mattysweeps** [comment]
Thank you @Aaron1011

**Comment by Aaron1011** [comment]
@m0pt0pmatt: I found the issue.

In 1.10, Method from Vanilla's `IInventory`, such as `clear`, are implemented in each `TileEntity` (e.g. `TileEntityChest`). Mixin will correctly see tyhat the `@Intrinsic` method already exists in the target, and skip adding it.

However, in 1.11, many of those methods (such as `clear`) were moved to `TileEntityLockableLoot`. Since our `@Intrinsic`s were never updated for this change, Mixin will always add them, since their target will never exist.

I'll push a fix for this shortly.
